dali/kernels/erase/erase_gpu_test.cu

```diff
@@ -0,0 +1,390 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include <cmath>
+#include <complex>
+#include <tuple>
+#include <vector>
+
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/erase/erase_gpu.h"
+#include "dali/kernels/scratch.h"
+#include "dali/pipeline/data/tensor_list.h"
+#include "dali/test/tensor_test_utils.h"
+#include "dali/test/test_tensors.h"
+
+#include "dali/kernels/erase/erase_cpu.h"
+
+#include "dali/core/cuda_event.h"
+
+namespace dali {
+namespace kernels {
+
+template <int ndim>
+void debug_print(TensorListView<StorageCPU, uint8_t, ndim> tlv, int height, int width) {
+  for (int y = 0; y < height; y++) {
+    for (int x = 0; x < width; x++) {
+      std::cout << std::setw(2) << int(*tlv[0](y, x)) % 100;
+    }
+    std::cout << endl;
+  }
+}
+
+void verify_regions(ivec<4> region_dims, ivec<4> sample_dims, ivec<4> expected_cover) {
+  auto cover = get_region_cover(region_dims, sample_dims);
+  EXPECT_EQ(cover, expected_cover);
+  int idx = 0;
+  for (int d0 = 0; d0 < expected_cover[0]; d0++) {
+    for (int d1 = 0; d1 < expected_cover[1]; d1++) {
+      for (int d2 = 0; d2 < expected_cover[2]; d2++) {
+        for (int d3 = 0; d3 < expected_cover[3]; d3++) {
+          auto regions_start = get_region_start(idx, region_dims, sample_dims);
+          ivec<4> region_start_expected = {d0 * region_dims[0], d1 * region_dims[1],
+                                           d2 * region_dims[2], d3 * region_dims[3]};
+          EXPECT_EQ(regions_start, region_start_expected);
+          idx++;
+        }
+      }
+    }
+  }
+}
+
+TEST(EraseGpuKernelTest, CheckUtils) {
+  {
+    ivec<4> region_dims = {2, 2, 32, 32};
+    ivec<4> sample_dims = {16, 8, 64, 64};
+    ivec<4> expected_cover = {8, 4, 2, 2};
+    verify_regions(region_dims, sample_dims, expected_cover);
+  }
+  {
+    ivec<4> region_dims = {2, 2, 32, 32};
+    ivec<4> sample_dims = {16, 8, 32, 64};
+    ivec<4> expected_cover = {8, 4, 1, 2};
+    verify_regions(region_dims, sample_dims, expected_cover);
+  }
+  {
+    ivec<4> region_dims = {2, 2, 32, 32};
+    ivec<4> sample_dims = {16, 8, 64, 32};
+    ivec<4> expected_cover = {8, 4, 2, 1};
+    verify_regions(region_dims, sample_dims, expected_cover);
+  }
+  {
+    ivec<4> region_dims = {2, 2, 32, 32};
+    ivec<4> sample_dims = {16, 8, 32, 32};
+    ivec<4> expected_cover = {8, 4, 1, 1};
+    verify_regions(region_dims, sample_dims, expected_cover);
+  }
+}
+
+template <typename T, int ndim, int channel_dim = -1>
+struct EraseGpuKernelTest :
```

Comment:
That's just inverts the workaround, instead of abusing the value-parametrized test and instantiating it several times with different types for appropriate sets of values, we would abuse the type-parametrized type instances and would need to loop over sets of values inside of them.

Not much of a difference, but I would need to rework a lot of this boilerplate and I don't think it's worth the effort.